src/msgs/staking/query/is_executor_eligible.rs

```diff
@@ -0,0 +1,125 @@
+use core::str;
+
+use super::*;
+
+#[cfg_attr(feature = "cosmwasm", cw_serde)]
+#[cfg_attr(not(feature = "cosmwasm"), derive(Serialize, Debug, PartialEq))]
+#[cfg_attr(not(feature = "cosmwasm"), serde(rename_all = "snake_case"))]
+pub struct Query {
+    pub data: Bytes,
+}
+
+impl Query {
+    fn generate_hash(dr_id: &str, chain_id: &str, contract_addr: &str) -> Hash {
+        hash([
+            "is_executor_eligible".as_bytes(),
+            dr_id.as_bytes(),
+            chain_id.as_bytes(),
+            contract_addr.as_bytes(),
+        ])
+    }
+
+    #[cfg(not(feature = "cosmwasm"))]
+    fn decode(&self) -> Result<Vec<u8>> {
+        use base64::{prelude::BASE64_STANDARD, Engine};
+
+        let decoded = BASE64_STANDARD.decode(&self.data)?;
+        Ok(decoded)
+    }
+
+    #[cfg(feature = "cosmwasm")]
+    fn decode(&self) -> Result<Vec<u8>> {
+        Ok(self.data.to_vec())
+    }
+
+    fn dr_id_hex(&self) -> Result<String> {
+        let decoded = self.decode()?;
+        Ok(str::from_utf8(&decoded[67..131]).unwrap().to_owned())
+    }
+
+    pub fn parts(&self) -> Result<([u8; 33], Hash, Vec<u8>)> {
+        let decoded = self.decode()?;
+        let public_key = hex::decode(&decoded[..66])?.try_into().expect("Invalid public key");
+        let dr_id = hex::decode(&decoded[67..131])?.try_into().expect("Invalid dr_id");
+        let proof = hex::decode(&decoded[132..])?;
+
+        Ok((public_key, dr_id, proof))
+    }
+}
+
+impl VerifySelf for Query {
+    type Extra = ();
+
+    fn proof(&self) -> Result<Vec<u8>> {
+        let decoded = self.decode()?;
+        Ok(hex::decode(&decoded[132..])?)
+    }
+
+    fn msg_hash(&self, chain_id: &str, contract_addr: &str, _: Self::Extra) -> Result<Hash> {
+        Ok(Self::generate_hash(&self.dr_id_hex()?, chain_id, contract_addr))
+    }
+}
+
+pub struct QueryFactory {
+    dr_id:      String,
+    public_key: String,
+    hash:       Hash,
+}
+
+impl QueryFactory {
+    pub fn get_hash(&self) -> &[u8] {
+        &self.hash
+    }
+
+    #[cfg(not(feature = "cosmwasm"))]
+    fn encode_data(data: &str) -> Bytes {
+        use base64::{prelude::BASE64_STANDARD, Engine};
+
+        BASE64_STANDARD.encode(data)
+    }
+
+    #[cfg(feature = "cosmwasm")]
+    fn encode_data(data: &str) -> Bytes {
+        use cosmwasm_std::Binary;
+        Binary(data.as_bytes().to_vec())
+    }
+
+    #[cfg(test)]
+    pub(crate) fn create_query(self, proof: Vec<u8>) -> Query {
+        let data = format!("{}:{}:{}", self.public_key, self.dr_id, proof.to_hex());
+
+        Query {
+            data: Self::encode_data(&data),
+        }
+    }
```

Comment:
Imo it would be better to not mix test helpers and actual code.

Reply:
@gluax It may be nice to sort of expose this `let data = format!("{}:{}:{}", self.public_key, self.dr_id, proof.to_hex());` in a function. The overlay node right now has no way to just get the raw data (+ base64 encoded) because it only gets the QueryMsg which i cannot easily extract the data property

Reply:
Oh! Maybe I should make the `create_query` return the `Query` type instead? You'd have to call `.into` on it, but it would let you access the `parts` method as well as the field itself, which has the encoded string.

Reply:
I went for changing `create_message` to return a tuple of the message and the base64.